fhir_zulip_nlp/fhir_zulip_nlp.py

```diff
@@ -379,21 +403,40 @@ def create_report_thread_length(
     return df_report
 
 
-def create_report_users(df: pd.DataFrame) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
-    """Report: Users
-    # TODO: Bugfix: Major bug; respondent/author counts are not all correct. This is because (i) threads are being
-       counted multiple times when multiple keywords are matched against them, and (ii) we are *only* counting messages
-       within threads that have keyword matches; not every message in every thread that has a keyword match for any
-       message.
-    # todo: Pending completion of 'streams' issue, change output `stream_id` -> `stream` / `stream_name`
-    # todo: Does it make sense to refactor this to start with users first, then drill down?"""
-    stream_id_name_map = {'179202': 'terminology'}
+def create_report_users(
+    df_all: pd.DataFrame, df_matches: pd.DataFrame
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    """Report: Users"""
+    # display_recipient: Data on the recipient of the message; either the name of a stream or a dictionary containing
+    # basic data on the users who received the message. See: https://zulip.com/api/get-messages
+    stream_name_field = 'display_recipient'
+
+    # Add only messages from `df_all` that aren't in `df_matches`
+    # - Create common primary key
+    for df_i in [df_all, df_matches]:
```

Comment:
Lots of changes in this function to deal w/ the two source datasets: (i) all messages (new), (ii) messages that match our keywords.

This would have been a bit easier if Zulip had the concept of / field for `thread` in their API. I ascertained this by combining `stream` + `subject`, because I believe you can have the same subject in multiple streams, but not duplicate subject names within a stream (at least I hope not).